opt: add PushSelectIntoWindow

justinj opened the pull request:
This commit adds a rule to push select filters into a Window expression.
This is valid to do whenever a filter is bound by the partition columns
of the Window expression.

Release note: None

Commit:
opt: add PushSelectIntoWindow

This commit adds a rule to push select filters into a Window expression.
This is valid to do whenever a filter is bound by the partition columns
of the Window expression.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37914)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-:-Lg4_lQiDBYY5gqII0QW:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/opt/norm/testdata/rules/window, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-Lg4_X5q3_L56jF7s1kS:-Lg4_X5q3_L56jF7s1kT:b-o4d3vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8e6b2e71fd05c4c0c5eabc06dbb1d513e96a2d8/pkg/sql/opt/norm/testdata/rules/window#L123)):*
> ```
> 
> # Only push down filters bound by the partition cols.
> ```

[nit] the comments should "stick" to the testcases, otherwise it's not immediately clear if they're talking about the one below or the one above


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-:-LgDBlTLEAK-bEqan8tn:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/opt/norm/testdata/rules/window, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-Lg4_X5q3_L56jF7s1kS:-LgDBmsq9eV1MnwLiqeH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8e6b2e71fd05c4c0c5eabc06dbb1d513e96a2d8/pkg/sql/opt/norm/testdata/rules/window#L123)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] the comments should "stick" to the testcases, otherwise it's not immediately clear if they're talking about the one below or the one above
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
opt: make PushSelectIntoWindow less restrictive

Immediately after bors'ing the previous PR about this rule I realized it
was too restrictive: we don't need to limit ourselves to pushing down
filters that are bound by the partition, we can push down any filters
that are bound by the *closure* of the partition.

Release note: None

justinj commented:
bors r-

justinj commented:
@RaduBerinde bors is broken right now, but right after bors'ing I realized that the rule as I wrote it at first was more restrictive than needed. I added a second commit which relaxes it. PTAL if you get a chance!

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-:-LgROpvl55sl-Jr0Y0OV:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @justinj, @rytaft, and @yuzefovich)

___
*[pkg/sql/opt/norm/testdata/rules/window, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-LgROk863efArZY7n6Kx:-LgROk863efArZY7n6Ky:bi50huv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea2d1cf13a5001c069474e57d639b46d7fb913b/pkg/sql/opt/norm/testdata/rules/window#L216)):*
> ```
>            └── s = 'foo' [type=bool, outer=(4), constraints=(/4: [/'foo' - /'foo']; tight), fd=()-->(4)]
> 
> # Can push down columns in the closure of the partition columns.
> ```

[nit] filters on columns


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-:-LgSwwy6D7nZwN9_oVqk:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/opt/norm/testdata/rules/window, line 216 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37914#-LgROk863efArZY7n6Kx:-LgSwvMZ6ktzELLLaXcx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eea2d1cf13a5001c069474e57d639b46d7fb913b/pkg/sql/opt/norm/testdata/rules/window#L216)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] filters on columns
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
